libkbfs/block_tree.go

```diff
@@ -291,159 +311,214 @@ func (bt *blockTree) getBlocksForOffsetRange(ctx context.Context,
 		// data to the end).
-		inRangeRight := endOff == nil || iptrOff.Less(endOff)
+		inRangeRight := job.endOff == nil || iptrOff.Less(job.endOff)
 		if !inRangeRight {
 			// This block is the first one past the offset range
 			// amount the children.
-			nextBlockOffsetThisLevel = iptrOff
-			break
+			results <- result{nextBlockOffset: iptrOff}
+			return nil
 		}
 
 		childPtr := info.BlockPointer
 		childIndex := i
-		respCh := make(chan resp, 1)
-		respChans = append(respChans, respCh)
-		// Don't reference the uncaptured `i` variable below.
-		eg.Go(func() error {
-			var pfr [][]parentBlockAndChildIndex
-			var blocks map[BlockPointer]Block
-			var nextBlockOffset Offset
-			// We only need to fetch direct blocks if we've been asked
-			// to do so.  If the direct type of the pointer is
-			// unknown, we can assume all the children are direct
-			// blocks, since there weren't multiple levels of
-			// indirection before the introduction of the flag.
-			if getDirect || childPtr.DirectType == IndirectBlock {
-				block, _, err := bt.getter(
-					groupCtx, bt.kmd, childPtr, bt.file, blockReadParallel)
-				if err != nil {
-					return err
-				}
 
-				// Recurse down to the level of the child.
-				pfr, blocks, nextBlockOffset, err = bt.getBlocksForOffsetRange(
-					groupCtx, childPtr, block, startOff, endOff, prefixOk,
-					getDirect)
-				if err != nil {
-					return err
-				}
-			} else {
-				// We don't care about direct blocks, so leave the
-				// `blocks` map `nil`.
-				pfr = [][]parentBlockAndChildIndex{nil}
-				nextBlockOffset = nil
+		childPath := make([]parentBlockAndChildIndex, len(job.pathPrefix)+1)
+		copy(childPath, job.pathPrefix)
+		childPath[len(childPath)-1] = parentBlockAndChildIndex{
+			pblock:     pblock,
+			childIndex: childIndex,
+		}
+
+		// We only need to fetch direct blocks if we've been asked
+		// to do so.  If the direct type of the pointer is
+		// unknown, we can assume all the children are direct
+		// blocks, since there weren't multiple levels of
+		// indirection before the introduction of the flag.
+		if job.getDirect || childPtr.DirectType == IndirectBlock {
+			// Create a task for the child, and either enqueue it or recurse to it.
+
+			subTask := task{
+				ptr:        childPtr,
+				pblock:     nil,
+				pathPrefix: childPath,
+				startOff:   job.startOff,
+				endOff:     job.endOff,
+				prefixOk:   job.prefixOk,
+				getDirect:  job.getDirect,
 			}
 
-			// Append self to the front of every path.
-			var r resp
-			for _, p := range pfr {
-				newPath := append([]parentBlockAndChildIndex{{
-					pblock:     pblock,
-					childIndex: childIndex,
-				}}, p...)
-				r.pathsFromRoot = append(r.pathsFromRoot, newPath)
+			// process the subtask by passing it to another goroutine; if they're all occupied, recurse instead.
+			select {
+			case jobs <- subTask:
+				wg.Add(1)
+			default:
+				bt.processGetBlocksTask(ctx, wg, subTask, jobs, results)
```

Comment:
I don't really get the benefit of re-using this worker to process the next job immediately, rather than making the job queue the same size as the number of workers, and letting this worker just pick the next thing from the queue.  That would make things easier to reason about, I think, especially when trying to inspect stack traces.

Reply:
The jobs queue would need to be arbitrarily large under this architecture to allow that - each job can queue lots more jobs, and so the queue can become full, leading to a deadlock where all the workers are trying to enqueue jobs. If we switch to the workerpool you linked in the top comment, we avoid this.

Reply:
Ah fair point.  During my review I didn't remember that this was inside of a loop over all the pointers in the block, rather than a single call (after which the worker exited and pulled a new job off the queue anyway).

But yeah, if you like workerpool, then it'd be nice to get rid of this, to make the code cleaner.  I think letting the job queue get as big as it needs to be is fine in this case.  I don't expect that to ever approach memory limits.